### PR TITLE
All configs on chain

### PR DESCRIPTION
**Pull request by vivianjeng**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

`numAttestationsPerProof` is not on-chain

Issue Number: N/A

## What is the new behavior?

- Use a struct `Config` in `Unirep.sol` to store all configs
- UserState can query all configs from contract instead of importing `NUM_ATTESTATIONS_PER_PROOF` from circuit config
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/contracts/contracts/Unirep.sol [hunk 1]
```diff
@@ -693,25 +663,26 @@ contract Unirep is IUnirep, zkSNARKHelper, Hasher, VerifySignature {
...
-        publicSignals[2 + numEpochKeyNoncePerEpoch] = input.blindedUserStates[
-            0
-        ];
-        publicSignals[3 + numEpochKeyNoncePerEpoch] = input.blindedUserStates[
-            1
-        ];
-        publicSignals[4 + numEpochKeyNoncePerEpoch] = input.fromGlobalStateTree;
-        for (uint8 i = 0; i < numEpochKeyNoncePerEpoch; i++) {
-            publicSignals[5 + numEpochKeyNoncePerEpoch + i] = input
+        publicSignals[1 + config.numEpochKeyNoncePerEpoch] = input
```
Comment (vuvoth) [inline packages/contracts/contracts/Unirep.sol:672]: We should "cache" config.numEpochKeyNoncePreEpoch and reuse it. Save gas I think. eg 
```
uint256 numberEpochKeyNoncePerEpoch = config.numEpochKeyNoncePerEpoch
```

Comment (vuvoth) [reply packages/contracts/contracts/Unirep.sol:672]: For read data from storage space, we use SLOAD of code. Around 100 gas/one time I think.
https://github.com/wolflo/evm-opcodes/blob/main/gas.md#a6-sload

File: packages/contracts/contracts/Unirep.sol [hunk 2]
```diff
@@ -23,25 +23,13 @@ contract Unirep is IUnirep, zkSNARKHelper, Hasher, VerifySignature {
     IVerifier internal reputationVerifier;
     IVerifier internal userSignUpVerifier;
 
-    uint256 public currentEpoch = 1;
-
-    uint256 public immutable epochLength;
+    Config public config;
```
Comment (vuvoth) [inline packages/contracts/contracts/Unirep.sol:26]: Config should be immutable right?

File: packages/contracts/contracts/Unirep.sol (hunk 1 above)
Comment (vuvoth) [reply packages/contracts/contracts/Unirep.sol:672]: Not sure how it apply on immutable data.

File: packages/contracts/package.json [hunk 3]
```diff
@@ -7,7 +7,7 @@
     "license": "MIT",
     "scripts": {
         "pre": "node scripts/prepare",
-        "build": "./scripts/buildVerifiers.sh && yarn compile && hardhat typechain && tsc",
+        "build": "npx ts-node ./scripts/buildVerifiers.ts && yarn compile && hardhat typechain && tsc",
```
Comment (vimwitch) [inline packages/contracts/package.json:10]: No need for `npx` here.

File: packages/core/cli/attest.ts [hunk 4]
```diff
@@ -76,7 +76,7 @@ const attest = async (args: any) => {
         args.contract,
         provider
     )
-    const attestingFee = await unirepContract.attestingFee()
+    const attestingFee = (await unirepContract.config()).attestingFee
```
Comment (vimwitch) [inline packages/core/cli/attest.ts:79]: Could destructure here

Comment (vimwitch) [reply packages/core/cli/attest.ts:79]: This same type syntax is used in lots of other places. It's fine to leave it as is

File: packages/contracts/contracts/Unirep.sol (hunk 2 above)
Comment (vivianjeng) [reply packages/contracts/contracts/Unirep.sol:25]: I don't think so 😅
```
TypeError: Immutable variables cannot have a non-value type.
  --> contracts/Unirep.sol:25:5:
   |
25 |     Config public immutable config;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by vimwitch — APPROVED** [review]

**Review by vuvoth — APPROVED** [review]

**Pull request merged**